### DISCUSSION TITLE
【踩坑】SyncStructure在mysql中不支持默认值

### DISCUSSION THREAD

**Discussion by shaukei** [dotnetcore/FreeSql#1369] [General]
昨天用freesql试了一下威力，果然很不错。可以提高code效率，也不用头痛底层业务逻辑（重连，SQL）。赞一个！

但是第一次用，也避免不了踩坑，发现了一些问题

按照文档，直接复制粘贴，
`.UseAutoSyncStructure(true) `
将上面这个也保留了，这到没什么，因为table class是用工具生成，相信问题不大
执行完一条数据插入时候，然后业务系统一直报警。
监控sql发现 有改表SQL
将原有的字段默认值删了。而保留了no null。。导致业务数据插不进表（这里可以认为表设计不合理，但是这种场景是确实存在的）

这里，其实是一个坑，我的理解是由于codefrist没有支持default，所以就会有以上的问题。

希望能有好方案支持好这个业务场景。

**Comment by densen2014** [comment]
请仔细阅读wiki, UseAutoSyncStructure不建议生产环境使用, 后面有三个叹号

"我的理解是由于codefrist没有支持default，所以就会有以上的问题。" 不是这样理解的

**Reply by shaukei** [reply]
Hi @densen2014 不是用在prod，用在开发环境，也是有这个问题，会把表的默认值给删除了，对吧。

或者我这样理解，这里是code first，所以是code来主导。
如果是这样的话，那必须code first的支持是完整的，例如需要把default也支持了。对吧。

还是数据表设计规范里面，有 “不应该设置字段默认值”

从开发角度来说，表设置默认值也是有场景的，甚至是 datetime | timestamp 字段类型，也会有on update current_timestamp
当然，这种特性可以在代码里面实现，但是在db实现这个功能，不是更加符合场景需求呢

我的意思是 code first没有问题，但是前提是需要支持好数据库特性，要不然 coder和dba要吵架了！

我是觉得，code first介绍没有问题，但是不应该在fsql基础功能usage code上出现。因为fsql新人认识得不够深入，容易变成踩坑！

**Reply by densen2014** [reply]
必须有足够的认识才能在生产环境里使用。已经写很清楚啦，我觉得这不叫“坑”了。都写那么明白还勇往直前冲这谁都挡不住啊。

**Reply by shaukei** [reply]
我也认同，需要足够认识才能用在生产！我也为我的大意付出了代价。我只是希望，优化文档！

我想说的，并不是要不要用在生产，而是 在文档里面，不要将 **不建议用在prod** 的代码放上去，
要放，其实是可以类似
`.UseAutoSyncStructure(**_false_**)`
这样安全一点的做法。
或者
![image](https://user-images.githubusercontent.com/1581783/205812975-e27094e0-25cc-445c-886f-8da0d6c11591.png)

加上该行代码不要用在生产的注释！

**Reply by densen2014** [reply]
我觉得你这就是在抬杠。


原文链接


https://github.com/dotnetcore/FreeSql/wiki/CodeFirst#freesql%E6%8F%90%E4%BE%9B%E4%B8%A4%E7%A7%8Dcodefirst%E7%A7%BB%E8%BF%81%E6%96%B9%E6%B3%95%E8%87%AA%E5%8A%A8%E5%92%8C%E6%89%8B%E5%8A%A8

FreeSql提供两种CodeFirst移迁方法，自动和手动。

注意：谨慎、谨慎、谨慎在生产环境中使用该功能。

注意：谨慎、谨慎、谨慎在生产环境中使用该功能。

注意：谨慎、谨慎、谨慎在生产环境中使用该功能。

![image](https://user-images.githubusercontent.com/8428709/205815277-43a1c2fe-0d6b-437b-8dd9-8974b8bf94b9.png)

**Reply by densen2014** [reply]
您提的意见很中肯。我们也完善一下文档，注意事项再最重标记一下，谢谢反馈！

**Reply by shaukei** [reply]
谢谢，我也是希望新人看文档时候可以及时注意到 重点提示！

**Comment by shaukei** [comment]
连建议都可以认作是抬杠，好吧，你对！

另外，fsql这个产品确实不错。fsql team，加油！
